src/NoiseTexture.cpp

```diff
@@ -14,90 +14,89 @@ using namespace graphics;
+		m_pTexture[i]->clampS = 1;
+		m_pTexture[i]->clampT = 1;
+		m_pTexture[i]->frameBufferTexture = CachedTexture::fbOneSample;
+		m_pTexture[i]->maskS = 0;
+		m_pTexture[i]->maskT = 0;
+		m_pTexture[i]->mirrorS = 0;
+		m_pTexture[i]->mirrorT = 0;
+		m_pTexture[i]->realWidth = 640;
+		m_pTexture[i]->realHeight = 580;
+		m_pTexture[i]->textureBytes = m_pTexture[i]->realWidth * m_pTexture[i]->realHeight;
+		textureCache().addFrameBufferTextureSize(m_pTexture[i]->textureBytes);
 
-	{
-		Context::InitTextureParams params;
-		params.handle = m_pTexture->name;
-		params.width = m_pTexture->realWidth;
-		params.height = m_pTexture->realHeight;
-		params.internalFormat = internalcolorFormat::RED;
-		params.format = colorFormat::RED;
-		params.dataType = datatype::UNSIGNED_BYTE;
-		gfxContext.init2DTexture(params);
-	}
-	{
-		Context::TexParameters params;
-		params.handle = m_pTexture->name;
-		params.target = textureTarget::TEXTURE_2D;
-		params.textureUnitIndex = textureIndices::NoiseTex;
-		params.minFilter = textureParameters::FILTER_NEAREST;
-		params.magFilter = textureParameters::FILTER_NEAREST;
-		gfxContext.setTextureParameters(params);
+		{
+			Context::InitTextureParams params;
+			params.handle = m_pTexture[i]->name;
+			params.width = m_pTexture[i]->realWidth;
+			params.height = m_pTexture[i]->realHeight;
+			params.internalFormat = internalcolorFormat::RED;
+			params.format = colorFormat::RED;
+			params.dataType = datatype::UNSIGNED_BYTE;
+			gfxContext.init2DTexture(params);
+		}
+		{
+			Context::TexParameters params;
+			params.handle = m_pTexture[i]->name;
+			params.target = textureTarget::TEXTURE_2D;
+			params.textureUnitIndex = textureIndices::NoiseTex;
+			params.minFilter = textureParameters::FILTER_NEAREST;
+			params.magFilter = textureParameters::FILTER_NEAREST;
+			gfxContext.setTextureParameters(params);
+		}
+		unsigned char ptr[m_pTexture[i]->textureBytes];
+		for (u32 y = 0; y < m_pTexture[i]->realHeight; ++y)     {
+			for (u32 x = 0; x < m_pTexture[i]->realWidth; ++x)
+				ptr[x + y*m_pTexture[i]->realWidth] = rand() & 0xFF;
+		}
+		{
+			Context::UpdateTextureDataParams params;
+			params.handle = m_pTexture[i]->name;
+			params.textureUnitIndex = textureIndices::NoiseTex;
+			params.width = 640;
+			params.height = 580;
+			params.format = colorFormat::RED;
+			params.dataType = datatype::UNSIGNED_BYTE;
+			params.data = ptr;
+			gfxContext.update2DTexture(params);
+		}
 	}
-
-	// Generate Pixel Buffer Object. Initialize it with max buffer size.
-	m_pbuf.reset(gfxContext.createPixelWriteBuffer(m_pTexture->textureBytes));
 }
 
 void NoiseTexture::destroy()
 {
-	textureCache().removeFrameBufferTexture(m_pTexture);
-	m_pTexture = nullptr;
-	m_pbuf.reset();
+	for (u32 i = 0; i < NOISE_TEX_NUM; ++i) {
+		textureCache().removeFrameBufferTexture(m_pTexture[i]);
+		m_pTexture[i] = nullptr;
+	}
 }
 
 void NoiseTexture::update()
 {
-	if (!m_pbuf || m_pTexture == nullptr)
+	int tex_num = rand() % NOISE_TEX_NUM;
```

Comment:
I don't think that rand will often return the same value, but when it does, you must not use the same texture twice.